libraries/AP_NavEKF3/AP_NavEKF3_core.cpp

```diff
@@ -47,34 +48,75 @@ bool NavEKF3_core::setup_core(NavEKF3 *_frontend, uint8_t _imu_index, uint8_t _c
     _ahrs = frontend->_ahrs;
 
     /*
-      the imu_buffer_length needs to cope with a 260ms delay at a
-      maximum fusion rate of 100Hz. Non-imu data coming in at faster
-      than 100Hz is downsampled. For 50Hz main loop rate we need a
-      shorter buffer.
+      The imu_buffer_length needs to cope with the worst case sensor delay at the
+      maximum fusion rate of 100Hz. Non-imu data coming in at faster than 100Hz is
+      downsampled. For 50Hz main loop rate we need a shorter buffer.
      */
-    if (_ahrs->get_ins().get_sample_rate() < 100) {
-        imu_buffer_length = 13;
+
+    // Calculate the expected EKF time step
+    if (_ahrs->get_ins().get_sample_rate() > 0) {
+        dtEkfAvg = 1.0f / _ahrs->get_ins().get_sample_rate();
+        dtEkfAvg = MAX(dtEkfAvg,EKF_TARGET_DT);
     } else {
-        // maximum 260 msec delay at 100 Hz fusion rate
-        imu_buffer_length = 26;
+        return false;
     }
-    if(!storedGPS.init(OBS_BUFFER_LENGTH)) {
+
+    // Wait up to 30 seconds for all GPS units to finish their configuration.
+    // Until this has occurred we will not know what type of GPS is being used
+    // and what its time delay is
+    if (!_ahrs->get_gps().all_configured() && (AP_HAL::millis() < 30E3)) {
+        if (AP_HAL::millis() - lastInitFailReport_ms > 5000) {
+            lastInitFailReport_ms = AP_HAL::millis();
+            GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "EKF3 waiting for GPS config data");
+        }
         return false;
+    } else if (!_ahrs->get_gps().all_configured()) {
```

Comment:
I can't see a reason to have this case. We should always succeed at configuring a GPS, unless the user requested to not configure it, in that case we should never opt out of the time delay, and instead should always be waiting until we succeed configuration.